File: backend/app/schemas/tag.py
"""Schemas de Tag: payloads de entrada e saída de
`app/api/routes/tag.py`.
"""
from pydantic import BaseModel, Field, field_validator

from app.schemas.base import OrmBaseModel

_PADRAO_COR_HEX = r"^#[0-9A-Fa-f]{6}$"


class TagCreate(BaseModel):
    nome: str = Field(min_length=1, max_length=60)
    cor: str | None = Field(default=None, pattern=_PADRAO_COR_HEX)

    @field_validator("nome", mode="before")
    @classmethod
    def _remover_espacos_nas_pontas(cls, nome: str) -> str:
        # mode="before": roda ANTES do min_length ser conferido, senao um
        # nome so de espacos (" ") passaria a checagem (tamanho 1) e so
        # viraria vazio depois - min_length nunca pegaria esse caso. Sem a
        # normalizacao, "viagem" e "viagem " tambem seriam nomes DIFERENTES
        # pro UniqueConstraint(usuario_id, nome), esvaziando na pratica a
        # regra de "nome unico por usuario".
        return nome.strip() if isinstance(nome, str) else nome


class TagUpdate(BaseModel):
    """Todos os campos opcionais - semântica de PATCH. `usuario_id` nunca
    aparece aqui: posse não se atualiza via payload."""

    nome: str | None = Field(default=None, min_length=1, max_length=60)
    cor: str | None = Field(default=None, pattern=_PADRAO_COR_HEX)
    ativo: bool | None = None

    @field_validator("nome", mode="before")
    @classmethod
    def _remover_espacos_nas_pontas(cls, nome: str | None) -> str | None:
        return nome.strip() if isinstance(nome, str) else nome


class TagRead(OrmBaseModel):
    id: int
    nome: str
    cor: str | None
    ativo: bool


class TagUso(BaseModel):
    """Resposta de `GET /tags/{id}/uso` - Etapa F10 (Exclusão definitiva,
    `docs/analise-arquitetural-exclusao.md`, seção 2.3). Só informativo:
    Tag nunca bloqueia exclusão por uso, este número existe só para o
    frontend avisar o usuário antes de confirmar."""

    transacoes_vinculadas: int

File: frontend/src/types/tag.ts
/**
 * Espelha 1:1 `app/schemas/tag.py` — conferido por leitura direta do
 * backend (`TagCreate`/`TagUpdate`/`TagRead`), não de documentação. Ver
 * docs/analise-arquitetural-tag-frontend.md, seção 2.1.
 *
 * Diferente de `CategoriaRead`, não há campo computado equivalente a
 * `e_do_sistema` — toda Tag pertence a um usuário, sem exceção.
 */

export interface TagRead {
  id: number;
  nome: string;
  cor: string | null;
  ativo: boolean;
}

export interface TagCreate {
  nome: string;
  cor?: string | null;
}

/** Semântica de PATCH no backend — todo campo omitido permanece intocado.
 * Inclui `ativo`: reativar uma tag desativada é um PATCH `{ ativo: true }`,
 * mesmo endpoint usado para editar qualquer outro campo (mesmo padrão de
 * `ContaUpdate`/`CategoriaUpdate`). */
export interface TagUpdate {
  nome?: string;
  cor?: string | null;
  ativo?: boolean;
}

/** Espelha `TagUso` (`app/schemas/tag.py`) — resposta de
 * `GET /tags/{id}/uso`, Etapa F10 (Exclusão definitiva,
 * `docs/analise-arquitetural-exclusao.md`, seção 2.3). Só informativo: Tag
 * nunca bloqueia exclusão por uso. */
export interface TagUso {
  transacoes_vinculadas: number;
}
